### PR TITLE
SAMZA-1146: TaskCallbackManager commit fix.

### PR DESCRIPTION
**Pull request by shanthoosh**
Each task callback in samza belongs to different SystemStreamPartition. When multiple callbacks in contagious order are available for commit, callback with highest sequence number is chosen for commit. This will prevent checkpointing of completed callbacks that has commit request and doesn't have highest sequence number. Upon task restart this will lead to duplicate reprocessing of already processed messages (since completed callbacks for some SystemStreamPartition's aren't committed earlier).

This PR fixes it and commits all completed callbacks that has commit request defined. Added a test to verify the behavior.

### CODE REVIEW

**Comment by shanthoosh** [comment]
@nickpan47 : Can you please take a look, when you get chance.

File: samza-core/src/main/java/org/apache/samza/task/TaskCallbackManager.java [hunk 1]
```diff
@@ -46,22 +49,22 @@
...
      */
-    TaskCallbackImpl update(TaskCallbackImpl cb) {
+    List<TaskCallbackImpl> update(TaskCallbackImpl cb) {
+
       synchronized (lock) {
         callbacks.add(cb);
-
         TaskCallbackImpl callback = null;
-        TaskCallbackImpl callbackToCommit = null;
+        List<TaskCallbackImpl> committableCallbacks = new ArrayList<>();
```
Comment (xinyuiscool) [inline samza-core/src/main/java/org/apache/samza/task/TaskCallbackManager.java]: Actually we don't need to have a separate commitableCallbacks. We need to update for all the callbacks below the low watermark. So this should just be callbacksToUpdate or something

File: samza-core/src/main/java/org/apache/samza/task/TaskCallbackManager.java [hunk 2]
```diff
@@ -46,22 +49,22 @@
...
         TaskCoordinator.RequestScope shutdownRequest = null;
         // look for the last contiguous callback
         while (!callbacks.isEmpty() && callbacks.peek().matchSeqNum(nextSeqNum)) {
           ++nextSeqNum;
           callback = callbacks.poll();
 
           if (callback.coordinator.commitRequest().isDefined()) {
```
Comment (xinyuiscool) [inline samza-core/src/main/java/org/apache/samza/task/TaskCallbackManager.java]: no need to check. Add the callback to the list.

File: samza-core/src/main/java/org/apache/samza/task/TaskCallbackManager.java [hunk 3]
```diff
@@ -70,16 +73,18 @@ TaskCallbackImpl update(TaskCallbackImpl cb) {
         }
 
         // if there is no manual commit, use the highest contiguous callback message offset
-        if (callbackToCommit == null) {
-          callbackToCommit = callback;
+        if (committableCallbacks.isEmpty() && callback != null) {
```
Comment (xinyuiscool) [inline samza-core/src/main/java/org/apache/samza/task/TaskCallbackManager.java]: no need to do this.

File: samza-core/src/main/java/org/apache/samza/task/TaskCallbackManager.java [hunk 4]
```diff
@@ -46,22 +49,22 @@
...
         while (!callbacks.isEmpty() && callbacks.peek().matchSeqNum(nextSeqNum)) {
           ++nextSeqNum;
           callback = callbacks.poll();
 
           if (callback.coordinator.commitRequest().isDefined()) {
-            callbackToCommit = callback;
+            committableCallbacks.add(callback);
           }
 
           if (callback.coordinator.shutdownRequest().isDefined()) {
```
Comment (xinyuiscool) [inline samza-core/src/main/java/org/apache/samza/task/TaskCallbackManager.java]: no need to do this anymore.

File: samza-core/src/main/java/org/apache/samza/task/TaskCallbackManager.java [hunk 5]
```diff
@@ -70,16 +73,18 @@ TaskCallbackImpl update(TaskCallbackImpl cb) {
...
 
         // if there is no manual commit, use the highest contiguous callback message offset
-        if (callbackToCommit == null) {
-          callbackToCommit = callback;
+        if (committableCallbacks.isEmpty() && callback != null) {
+          committableCallbacks.add(callback);
         }
 
         // if there is a shutdown request, merge it into the coordinator to commit
         if (shutdownRequest != null) {
```
Comment (xinyuiscool) [inline samza-core/src/main/java/org/apache/samza/task/TaskCallbackManager.java]: no need to do this.

File: samza-core/src/main/java/org/apache/samza/task/TaskCallbackManager.java [hunk 6]
```diff
@@ -46,22 +49,22 @@
...
         callbacks.add(cb);
-
         TaskCallbackImpl callback = null;
-        TaskCallbackImpl callbackToCommit = null;
+        List<TaskCallbackImpl> committableCallbacks = new ArrayList<>();
         TaskCoordinator.RequestScope shutdownRequest = null;
         // look for the last contiguous callback
         while (!callbacks.isEmpty() && callbacks.peek().matchSeqNum(nextSeqNum)) {
           ++nextSeqNum;
           callback = callbacks.poll();
```
Comment (xinyuiscool) [inline samza-core/src/main/java/org/apache/samza/task/TaskCallbackManager.java]: Seems to me the logic in this while loop should be extremely simple now. Just do 

callbacksToUpdate.add(callback)

and that should be it. Please think about it.

File: samza-core/src/main/java/org/apache/samza/task/TaskCallbackManager.java (hunk 1 above)
Comment (shanthoosh) [inline samza-core/src/main/java/org/apache/samza/task/TaskCallbackManager.java]: Ranemed to callbacksToUpdate.

File: samza-core/src/main/java/org/apache/samza/task/TaskCallbackManager.java (hunk 6 above)
Comment (shanthoosh) [inline samza-core/src/main/java/org/apache/samza/task/TaskCallbackManager.java]: Yes, this is fixed. 
I'd misunderstanding that task.commit is for commit in only the callback  SSP that is processed & committed by the task. Thanks for pointing it out that it's for all the input SSP's that are processed until the callback SSP that's processed & committed.

File: samza-core/src/main/java/org/apache/samza/task/TaskCallbackManager.java (hunk 2 above)
Comment (shanthoosh) [inline samza-core/src/main/java/org/apache/samza/task/TaskCallbackManager.java]: Removed.

File: samza-core/src/main/java/org/apache/samza/task/TaskCallbackManager.java (hunk 4 above)
Comment (shanthoosh) [inline samza-core/src/main/java/org/apache/samza/task/TaskCallbackManager.java]: Removed.

File: samza-core/src/main/java/org/apache/samza/task/TaskCallbackManager.java (hunk 5 above)
Comment (shanthoosh) [inline samza-core/src/main/java/org/apache/samza/task/TaskCallbackManager.java]: Removed.

**Comment by prateekm** [comment]
@shanthoosh Just FYI, you should rebase from upstream master before updating the PR so that upstream commits don't show up as part of your changes.

**Comment by shanthoosh** [comment]
@prateekm : 

I did a bad rebase because of which it showed up as a part of my PR.  I will be careful from now on.

**Pull request closed without merging**